### PR TITLE
[lxd][runtime] Add LXD Runtime

### PR DESCRIPTION
**Pull request by arif-ali**
---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [ ] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?

### CODE REVIEW

File: sos/policies/runtimes/lxd.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+    def get_containers(self, get_all=False):
+        """Get a list of containers present on the system.
+
+        :param get_all: If set, include stopped containers as well
+        :type get_all: ``bool``
+        """
+        containers = []
+
+        _cmd = "%s list --format json" % (self.binary)
```
Comment (TurboTurtle) [inline sos/policies/runtimes/lxd.py:41]: Nitpick - the rest of the runtime uses f-strings (yay!), so let's use one here to for consistency, and for getting away from string substitution.

File: sos/policies/runtimes/lxd.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    binary = 'lxc'
+
+    def check_is_active(self, sysroot=None):
+        # the daemon must be running
+        if (is_executable('lxc', sysroot) and
+                (self.policy.init_system.is_running('lxd') or
+                 self.policy.init_system.is_running('snap.lxd.daemon'))):
+            self.active = True
+            return True
+        return False
```
Comment (TurboTurtle) [inline sos/policies/runtimes/lxd.py:31]: Use `self.policy.sysroot` here, as the only place this is called from Policy doesn't pass a sysroot. On reviewing this, we should make a similar change to the `docker` runtime (in a different PR, no need to add that here).

File: sos/policies/runtimes/lxd.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            )
+            if out['status'] == 0:
+                out_json = json.loads(out['output'])
+                for profile in out_json:
+                    if profile['name'] == 'default':
+                        stg_pool = profile['devices']['root']['pool']
+                        break
+
+            out = sos_get_command_output(
+                f"{self.binary} storage volume list {stg_pool} --format json",
```
Comment (TurboTurtle) [inline sos/policies/runtimes/lxd.py:100]: We should have a default value for `stg_pool`, or raise an exception after the check if we don't find one. Like CodeQL mentions, there's a possibility here for `stg_pool` to be undefined when we make this call to `sos_get_command_output()`

**Review by TurboTurtle — COMMENTED** [review]
Looks mostly solid so far, just a couple minor comments. I'll get a PR up for the `docker` change shortly to cover that side of things.

File: sos/policies/runtimes/lxd.py (hunk 3 above)
Comment (pmoravec) [reply sos/policies/runtimes/lxd.py:100]: Indeed, our CI tests even detected it :) https://cirrus-ci.com/task/5971844470145024

File: sos/policies/runtimes/lxd.py (hunk 1 above)
Comment (arif-ali) [reply sos/policies/runtimes/lxd.py]: nice catch, I originally wrote this 5 months ago, so converted all to f-string, missed this one ;)

**Comment by arif-ali** [comment]
I'm not really happy with the way `get_volumes()` function is expecting the return, as `lxc` doesn't necessarily give ids, but names of the volumes. So, at the moment this is giving a list of names of volumes in the `default` storage pool or the pool from the `default` profile. This is going to be as good as it goes for the time being.

If this would have a refactor later, we could do with pool + name/id tuple, then that would work better for `lxd`, but that would mean that `docker`/`podman` ones would need to change too. Different concepts of volumes/storage, need to work on understanding how this is used in the wider `sos` aspect

**Review by TurboTurtle — APPROVED** [review]

**Review by pmoravec — APPROVED** [review]

**Pull request merged**